crates/cow-amm/src/amm.rs

```diff
@@ -0,0 +1,155 @@
+use {
+    anyhow::Result,
+    app_data::AppDataHash,
+    contracts::CowAmmLegacyHelper,
+    ethcontract::{Address, Bytes, U256},
+    model::{
+        interaction::InteractionData,
+        order::{BuyTokenDestination, OrderData, OrderKind, SellTokenSource},
+        signature::Signature,
+    },
+};
+
+#[derive(Clone, Debug)]
+pub struct Amm {
+    helper: contracts::CowAmmLegacyHelper,
+    address: Address,
+    tradeable_tokens: Vec<Address>,
+}
+
+impl Amm {
+    pub(crate) async fn new(address: Address, helper: &CowAmmLegacyHelper) -> Result<Self> {
+        let tradeable_tokens = helper.tokens(address).call().await?;
+
+        Ok(Self {
+            helper: helper.clone(),
+            address,
+            tradeable_tokens,
+        })
+    }
+
+    pub fn address(&self) -> &Address {
+        &self.address
+    }
+
+    /// Returns all tokens traded by this pool in stable order.
+    pub fn traded_tokens(&self) -> &[Address] {
+        &self.tradeable_tokens
+    }
+
+    /// Returns an order to rebalance the AMM based on the provided reference
+    /// prices. `prices` need to be computed using a common denominator and
+    /// need to be supplied in the same order as `traded_tokens` returns
+    /// token addresses.
+    pub async fn template_order(
+        &self,
+        prices: Vec<U256>,
+    ) -> Result<(
+        OrderData,
+        Signature,
+        Vec<InteractionData>,
+        Vec<InteractionData>,
+    )> {
+        let (order, pre_interactions, post_interactions, signature) =
+            self.helper.order(self.address, prices).call().await?;
+        self.convert_orders_reponse(order, signature, pre_interactions, post_interactions)
+    }
+
+    /// Converts a successful response of the CowAmmHelper into domain types.
+    /// Can be used for any contract that correctly implements the CoW AMM
+    /// helper interface.
+    fn convert_orders_reponse(
+        &self,
+        order: RawOrder,
+        signature: Bytes<Vec<u8>>,
+        pre_interactions: Vec<RawInteraction>,
+        post_interactions: Vec<RawInteraction>,
+    ) -> Result<(
```

Comment:
nit: what about returning a struct containing all the fields? big tuples are difficult to track, and easy to mix (e.i. from the signature of the function it is not known which one is pre_interactions and which one post_interactions)